File: src/gestorAplicacion/administracion/Vacuna.java
/* Autores: Diego Andres Gracia Granados, Daniel Giraldo Vanegas, Elian David Velandia Riveros, Juan Camilo Gutierrez Martinez
y Santiago Arboleda Acevedo */
package gestorAplicacion.administracion;

import gestorAplicacion.personas.Paciente;
import gestorAplicacion.servicios.CitaVacuna;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Objects;

//Clase destinada a los objetos vacuna
public class Vacuna implements Serializable {

    //Atributos
    private String tipo;
    private String nombre;
    private double precio;
    private ArrayList<String> tipoEps= new ArrayList<String>();
    private ArrayList<CitaVacuna> agenda= new ArrayList<CitaVacuna>();

    //Constructor
    public Vacuna(String tipo, String nombre, ArrayList<String> tipoEps, double precio) {
        this.tipo = tipo;
        this.nombre = nombre;
        this.tipoEps = tipoEps;
        this.precio=precio;
        //Se crea una agenda por defecto para cada vacuna
        agenda.add(new CitaVacuna("3 de Abril, 8:00 am", null, this));
        agenda.add(new CitaVacuna("3 de Abril, 11:00 am", null, this));
        agenda.add(new CitaVacuna("4 de Abril, 3:00 pm", null, this));
        agenda.add(new CitaVacuna("5 de Abril, 10:00 am", null, this));
    }

    //Métodos

    //Método que muestra las citas que no tienen paciente asignado de una vacuna
    public ArrayList<CitaVacuna> mostrarAgendaDisponible(){
        ArrayList<CitaVacuna> agendaDisponible= new ArrayList<CitaVacuna>();
        for (int i=1; i<=agenda.size();i++){
            if (agenda.get(i-1).getPaciente()==null){
                agendaDisponible.add(agenda.get(i-1));
            }
        }
        return agendaDisponible;
    }

    //Este método asigna un paciente a una cita de una vacuna
    public CitaVacuna actualizarAgenda(Paciente pacienteAsignado, byte numeroCita, ArrayList<CitaVacuna> agendaDisponible){
        CitaVacuna citaAsignada= null;
        for (int i=1 ; i<=agenda.size();i++){
            if(Objects.equals(agenda.get(i - 1).getFecha(), agendaDisponible.get(numeroCita - 1).getFecha())){
                agenda.get(i-1).setPaciente(pacienteAsignado);
                citaAsignada= agenda.get(i-1);
            }
        }
        return citaAsignada;
    }

    //Getters y Setters
    public String getTipo() {
        return tipo;
    }

    public ArrayList<String> getTipoEps() {
        return tipoEps;
    }

    public String getNombre() {
        return nombre;
    }

    public double getPrecio() {
        return precio;
    }

    public ArrayList<CitaVacuna> getAgenda() {
        return agenda;
    }

}

File: Python/src/gestor_aplicacion/administracion/vacuna.py
# Autores: Diego Andres Gracia Granados, Daniel Giraldo Vanegas, Elian David Velandia Riveros, Juan Camilo Gutierrez Martinez
# y Santiago Arboleda Acevedo


from src.gestor_aplicacion.servicios.cita_vacuna import CitaVacuna
from src.manejo_errores.error_aplicacion import SinAgenda

# Clase destinada a los objetos vacuna
class Vacuna:

    # Atributos y constructor
    def __init__(self, tipo, nombre, tipo_eps, precio):
        self._tipo = tipo
        self._nombre = nombre
        self._precio = precio
        self._tipo_eps = tipo_eps
        self._agenda = [
            CitaVacuna("9 de Enero, 8:00 am", None, self),
            CitaVacuna("25 de Junio, 10:00 am", None, self),
            CitaVacuna("31 de Diciembre, 3:00 pm", None, self),
            CitaVacuna("24 de Diciembre, 6:00 am", None, self)
        ]

    # Metodos

    # Método que muestra las citas que no tienen paciente asignado de una vacuna
    # No recibe ningun parametro y devuelve una lista de objetos Cita_Vacuna
    def mostrar_agenda_disponible(self):
        agenda_disponible = []
        for cita in self._agenda:
            if cita.paciente is None:
                agenda_disponible.append(cita)
        if len(agenda_disponible) != 0:
            return agenda_disponible
        else:
            raise SinAgenda()

    # Este método asigna un paciente a una cita de una vacuna
    # recibe un objeto de Paciente, un entero y una lista de Cita_Vacuna
    def actualizar_agenda(self, paciente_asignado, numero_cita, agenda_disponible):
        cita_asignada = None
        for cita in self._agenda:
            if cita.fecha == agenda_disponible[numero_cita - 1].fecha:
                cita.paciente=paciente_asignado
                cita_asignada = cita
        return cita_asignada

    @property
    def tipo(self):
        return self._tipo

    @property
    def tipo_eps(self):
        return self._tipo_eps

    @property
    def nombre(self):
        return self._nombre

    @property
    def precio(self):
        return self._precio

    @property
    def agenda(self):
        return self._agenda
